mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,398 @@
+
+                    if this.buf.len() < trailer_length {
+                        // Can't read trailer right now
+                        return Ok(());
+                    } else {
+                        let mut trailer_bytes = this.buf.take(trailer_length);
+                        trailer_bytes.put_u8(b'\n');
+
+                        let mut trailers_buf = [EMPTY_HEADER; 64];
+                        let parsed_trailers = match httparse::parse_headers(&trailer_bytes, &mut trailers_buf)
+                            .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?
+                        {
+                            Status::Complete((_, headers)) => Ok(headers),
+                            Status::Partial => Err(PostGrpcWebErr::InvalidRequest("Invalid".to_string())),
+                        }?;
+
+                        let mut trailers = HeaderMap::with_capacity(parsed_trailers.len());
+
+                        for parsed_trailer in parsed_trailers {
+                            let header_name = HeaderName::from_bytes(parsed_trailer.name.as_bytes())
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            let header_value = HeaderValue::from_bytes(parsed_trailer.value)
+                                .map_err(|err| PostGrpcWebErr::InvalidRequest(err.to_string()))?;
+                            trailers.insert(header_name, header_value);
+                        }
+
+                        *this.trailer = Some(trailers);
+
+                        *this.state = ReadState::Done;
+                    }
+                },
+                ReadState::Done => return Ok(()),
+            }
+        }
+    }
+}
+
+impl Body for ResponseBody {
+    type Data = Bytes;
+
+    type Error = PostGrpcWebErr;
+
+    fn poll_data(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        // Check if there's already some data in buffer and return that
+        if self.data.is_some() {
+            return Poll::Ready(self.data.take().map(|d| Ok(d.freeze())));
+        }
+
+        // If reading data is finished return `None`
+        if self.state.finished_data() {
+            return Poll::Ready(None);
+        }
+
+        loop {
+            // Read bytes from stream
+            if let Err(e) = ready!(self.as_mut().read_stream(cx)) {
+                return Poll::Ready(Some(Err(e)));
+            }
+
+            // Step the state machine
+            if let Err(e) = self.as_mut().step() {
+                return Poll::Ready(Some(Err(e)));
+            }
+
+            if self.data.is_some() {
+                // If data is available in buffer, return that
+                return Poll::Ready(self.data.take().map(|d| Ok(d.freeze())));
+            } else if self.state.finished_data() {
+                // If we finished reading data continue return `None`
+                return Poll::Ready(None);
```

Comment:
The proposal was to replace *both* `if self.data.is_some()` and `if self.state.finished_data()` with
```rust
// If reading data is finished return it
if self.state.finished_data() {
    return Poll::Ready(self.data.take().map(|data| Ok(data.freeze())));
}
```

If we finished reading data, the replacement code covers both `Some` and `None` cases.

Reply:
but I have a question @artemii235 , Why not just return `None` explicitly if the `self.state.finished_data()` instead of trying to take `self.data` again

Reply:
I propose the following diff. Here, we have fewer conditions, so the code is not `trying to take self.data again`. Both `Some` and `None` cases are covered by a single line.

```diff
index 6c45d809a..666661c5a 100644
--- a/mm2src/mm2_net/src/wasm/body_stream.rs
+++ b/mm2src/mm2_net/src/wasm/body_stream.rs
@@ -288,14 +288,8 @@ impl Body for ResponseBody {
     type Error = PostGrpcWebErr;
 
     fn poll_data(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Result<Self::Data, Self::Error>>> {
-        // Check if there's already some data in buffer and return that
-        if self.data.is_some() {
-            return Poll::Ready(self.data.take().map(|d| Ok(d.freeze())));
-        }
-
-        // If reading data is finished return `None`
         if self.state.finished_data() {
-            return Poll::Ready(None);
+            return Poll::Ready(self.data.take().map(|data| Ok(data.freeze())));
         }
 
         loop {
@@ -309,12 +303,8 @@ impl Body for ResponseBody {
                 return Poll::Ready(Some(Err(e)));
             }
 
-            if self.data.is_some() {
-                // If data is available in buffer, return that
-                return Poll::Ready(self.data.take().map(|d| Ok(d.freeze())));
-            } else if self.state.finished_data() {
-                // If we finished reading data continue return `None`
-                return Poll::Ready(None);
+            if self.state.finished_data() {
+                return Poll::Ready(self.data.take().map(|data| Ok(data.freeze())));
             } else if self.finished_stream {
                 // If stream is finished but data is not finished return error
                 return Poll::Ready(Some(Err(PostGrpcWebErr::InvalidRequest("Bad response".to_string()))));
```

Reply:
I see. thank you